sycl/source/detail/program_manager/program_manager.hpp

```diff
@@ -272,6 +280,10 @@ class ProgramManager {
   /// Access must be guarded by the \ref Sync::getGlobalLock()
   std::unordered_map<OSModuleHandle, KernelSetId> m_OSModuleKernelSets;
 
+  /// Maps names of kernels to their unique kernel IDs.
+  /// Access must be guarded by the \ref Sync::getGlobalLock()
+  std::unordered_map<std::string, kernel_id> m_KernelIDs;
```

Comment:
Wouldn't  `std:unordered_set` with https://github.com/intel/llvm/blob/sycl/sycl/source/detail/kernel_id_impl.hpp#L16 as a predicate be more efficient?

Reply:
I have made the change to std::unordered_set as I believe you are right that it is potentially faster. This also simplifies the implementation of `sycl::get_kernel_ids()`. On the other hand it does also complicate lookups as `find` now requires a `kernel_id` object, though as we hash by name it is able to find the correct `kernel_id` even if they are not equal according to `operator==`.
With C++20 this can be made nicer with [transparent hash and key-equal functions](https://en.cppreference.com/w/cpp/container/unordered_set/find). Would it make sense to make a note of that?

>  with https://github.com/intel/llvm/blob/sycl/sycl/source/detail/kernel_id_impl.hpp#L16

Seems to me like it should be using `EqualByNameComp`, but maybe I am missing something?

Reply:
So, now I'm not sure suggesting `std::unordered_set` was a good idea. Because I believe we should care more about lookup rather than initial insertion, but with `std::unordered_set` we really slowdown lookup by doing a couple of memory allocations.
Not sure if transparent keys will help here, because I would expect that `find(x)` has to iterate over all the values in the set and compare with each value. Maybe I'm wrong though, but quick googling didn't help.

Reply:
I agree that `std::unordered_set` might not be the right structure at the moment. However, I believe the transparent keys are meant for doing efficient lookup as the hasher must be transparent as well. However, that requires the runtime library to use C++20 so it is more suitable for a future change. I have added a note to do it when C++20 is enabled.